pkg/releasepayload/jobrunresult/helpers.go

```diff
@@ -16,3 +19,129 @@ func (in ByCoordinatesName) Len() int {
 func (in ByCoordinatesName) Swap(i, j int) {
 	in[i], in[j] = in[j], in[i]
 }
+
+func SetJobRunResult(results *[]v1alpha1.JobRunResult, newResult v1alpha1.JobRunResult) {
+	if results == nil {
+		results = &[]v1alpha1.JobRunResult{}
+	}
+	existingResult := FindJobRunResult(*results, newResult.Coordinates)
+	if existingResult == nil {
+		*results = append(*results, newResult)
+		return
+	}
+
+	existingResult.CompletionTime = newResult.CompletionTime
+	existingResult.HumanProwResultsURL = newResult.HumanProwResultsURL
+	existingResult.StartTime = newResult.StartTime
+	existingResult.State = newResult.State
+}
+
+func RemoveJobRunResult(results *[]v1alpha1.JobRunResult, coordinates v1alpha1.JobRunCoordinates) {
+	if results == nil {
+		results = &[]v1alpha1.JobRunResult{}
+	}
+	newResults := []v1alpha1.JobRunResult{}
+	for _, result := range *results {
+		if !cmp.Equal(result.Coordinates, coordinates) {
+			newResults = append(newResults, result)
+		}
+	}
+
+	*results = newResults
+}
+
+func FindJobRunResult(results []v1alpha1.JobRunResult, coordinates v1alpha1.JobRunCoordinates) *v1alpha1.JobRunResult {
+	for i := range results {
+		if cmp.Equal(results[i].Coordinates, coordinates) {
+			return &results[i]
+		}
+	}
+
+	return nil
+}
+
+func ComputeAggregateState(job v1alpha1.JobStatus) v1alpha1.JobState {
+	attempts := len(job.JobRunResults)
+
+	// If there are no results, then there is nothing to do...
+	if attempts == 0 {
+		return v1alpha1.JobStateUnknown
+	}
+
+	switch {
+	case job.AnalysisJobCount > 0:
+		return processAnalysisJobResults(job.JobRunResults, job.AnalysisJobCount)
+	case job.MaxRetries > 0:
+		return processRetryJobResults(job.JobRunResults, job.MaxRetries)
+	default:
+		return processJobResults(job.JobRunResults)
+	}
+}
+
+func interrogateJobResults(results []v1alpha1.JobRunResult) (pendingJobs, successfulJobs, failedJobs, unknownJobs []v1alpha1.JobRunResult) {
+	for _, result := range results {
+		switch result.State {
+		case v1alpha1.JobRunStatePending, v1alpha1.JobRunStateTriggered:
+			pendingJobs = append(pendingJobs, result)
+		case v1alpha1.JobRunStateSuccess:
+			successfulJobs = append(successfulJobs, result)
+		case v1alpha1.JobRunStateFailure, v1alpha1.JobRunStateAborted, v1alpha1.JobRunStateError:
+			failedJobs = append(failedJobs, result)
+		default:
+			unknownJobs = append(unknownJobs, result)
+		}
+	}
+
+	return pendingJobs, successfulJobs, failedJobs, unknownJobs
+}
+
+func processJobResults(results []v1alpha1.JobRunResult) v1alpha1.JobState {
```

Comment:
you need to doc this.  I didn't know what it did.

Reply:
Done